File: frontend/src/services/server-management.ts
import { authenticatedFetch } from "./auth";
import { requestError } from "./request-error";
import type { ServerSummary } from "../domain/inventory";

export interface ServerSettings {
  name: string;
  ip_address: string | null;
  ip_address_v6: string | null;
  domain: string | null;
  domain_v6: string | null;
  ipv6_enabled: boolean;
}
export interface SettingsResponse {
  server: ServerSummary;
  revision: string;
  updated_node_ids: string[];
  license_required: false;
}
export interface RemovalPreview {
  server_id: string;
  server_name: string;
  revision: string;
  nodes: { id: string; name: string }[];
  plans: { id: string; name: string }[];
  change_sets: { id: string; name: string }[];
  certificates: { id: string; name: string }[];
  command_count: number;
  unfinished_command_count: number;
  telemetry_count: number;
  user_count: number;
  blockers: string[];
}
const base = import.meta.env.VITE_API_BASE_URL ?? "";

async function request<T>(id: string, path: string, init?: RequestInit, fetcher = authenticatedFetch): Promise<T> {
  const response = await fetcher(`${base}/api/v1/servers/${id}/${path}`, init);
  if (!response.ok) {
    const error = await response.json().catch(() => null);
    throw requestError(error?.detail, `服务器请求失败（${response.status}）`);
  }
  return response.json() as Promise<T>;
}
export function getServerSettings(id: string, fetcher = authenticatedFetch) {
  return request<SettingsResponse>(id, "settings", undefined, fetcher);
}
export function updateServerSettings(id: string, settings: ServerSettings, revision: string, syncHosts: boolean, fetcher = authenticatedFetch) {
  return request<SettingsResponse>(id, "settings", { method: "PUT", headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ ...settings, expected_revision: revision, sync_node_hosts: syncHosts }) }, fetcher);
}
export function getServerRemoval(id: string, fetcher = authenticatedFetch) {
  return request<RemovalPreview>(id, "removal", undefined, fetcher);
}
export function removeServer(id: string, preview: RemovalPreview, name: string, fetcher = authenticatedFetch) {
  return request<{ server_id: string; removed_node_count: number; updated_plan_count: number }>(id, "remove", {
    method: "POST", headers: { "Content-Type": "application/json" }, body: JSON.stringify({
      expected_revision: preview.revision, confirm_name: name, acknowledge_remote_runtime: true,
    }),
  }, fetcher);
}

File: backend/app/open_node/domain/server_management.py
import re
from ipaddress import ip_address
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field, field_validator

from open_node.domain.inventory import ServerRead, _strip_required_text


class ServerSettings(BaseModel):
    model_config = {"extra": "forbid"}

    name: str = Field(min_length=1, max_length=120)
    ip_address: str | None = Field(default=None, max_length=255)
    ip_address_v6: str | None = Field(default=None, max_length=255)
    domain: str | None = Field(default=None, max_length=255)
    domain_v6: str | None = Field(default=None, max_length=255)
    ipv6_enabled: bool = False

    @field_validator("name")
    @classmethod
    def name_text(cls, value):
        return _strip_required_text(value, "server name")

    @field_validator("ip_address", "ip_address_v6")
    @classmethod
    def address(cls, value, info):
        if not value or not value.strip():
            return None
        address = ip_address(value.strip())
        if address.version != (6 if info.field_name.endswith("_v6") else 4):
            raise ValueError("IP address family does not match the field")
        return str(address)

    @field_validator("domain", "domain_v6")
    @classmethod
    def hostname(cls, value):
        if not value or not value.strip():
            return None
        try:
            normalized = value.strip().rstrip(".").encode("idna").decode("ascii").lower()
        except UnicodeError as exc:
            raise ValueError("Invalid hostname") from exc
        label = r"[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?"
        if len(normalized) > 253 or not re.fullmatch(rf"{label}(?:\.{label})*", normalized):
            raise ValueError("Use a hostname without a scheme, port or path")
        return normalized


class ServerSettingsUpdate(ServerSettings):
    expected_revision: str = Field(pattern=r"^[a-f0-9]{64}$")
    sync_node_hosts: bool = True


class ServerSettingsResponse(BaseModel):
    server: ServerRead
    revision: str
    updated_node_ids: list[UUID] = Field(default_factory=list)
    license_required: Literal[False] = False


class RemovalItem(BaseModel):
    id: str
    name: str


class ServerRemovalPreview(BaseModel):
    server_id: UUID
    server_name: str
    revision: str
    nodes: list[RemovalItem]
    plans: list[RemovalItem]
    change_sets: list[RemovalItem]
    certificates: list[RemovalItem]
    command_count: int
    unfinished_command_count: int
    telemetry_count: int
    user_count: int
    blockers: list[str]
    license_required: Literal[False] = False


class ServerRemovalRequest(BaseModel):
    model_config = {"extra": "forbid"}

    expected_revision: str = Field(pattern=r"^[a-f0-9]{64}$")
    confirm_name: str = Field(min_length=1, max_length=120)
    acknowledge_remote_runtime: Literal[True]


class ServerRemovalResponse(BaseModel):
    server_id: UUID
    removed_node_count: int
    updated_plan_count: int
    license_required: Literal[False] = False
